Prevent auto apply filters on slow dashboards

WiNloSt opened the pull request:
Epic: https://github.com/metabase/metabase/issues/29267
Product doc: https://www.notion.so/metabase/Prevent-auto-apply-filters-in-slow-dashboards-cc02d90dac9e42ee91f34c2a9810212f

This is 2/last of the epic

### Description

Allow users to prevent auto-apply filters on slow dashboard. Otherwise, each filter change could cause a dashboard refresh which could last a long time.

### How to verify

1. Create a dashboard
2. Add a question or model
3. Create a parameter and connect it to the entity added in step 2
4. Modify the filter added in step 3, the dashboard should refresh automatically
5. Open dashboard sidebar and toggle **Auto-apply filters**
6. Modify the filter, and the **Apply** button should appear, the dashboard should not refresh.
7. Click **Apply** button, and the dashboard should refresh.
8. Test the same dashboard in public dashboard, signed embed, and full-app embedding.

### Demo

https://www.loom.com/share/d08afee5fe9340dc9ed385e7c35918f7

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29569)
<!-- Reviewable:end -->

WiNloSt commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14378,6 +14378,22 @@ databaseChangeLog:
                   constraints:
                     nullable: false
 
+  - changeSet:
```
The BE change isn't meant to be merged, this will be here until it's properly implemented by BE team.

WiNloSt replied:
So @dpsutton implemented the BE portion in this BE, now this code is relevant and would be merged to `master`.

WiNloSt committed:
Only apply filters when clicking apply button

WiNloSt committed:
Add E2E test

WiNloSt committed:
Fix dashboard double fetch cards onload with parameters query

WiNloSt commented on frontend/src/metabase/core/components/Toggle/Toggle.tsx:
```diff
@@ -27,6 +28,26 @@ const Toggle = forwardRef(function Toggle(
     [onChange],
   );
 
+  if (label) {
```
It's worth noting that we have `<FormToggle />` which is normally used when we use Formik forms, and is more sophisticated. But it renders differently 
<img width="645" alt="image" src="https://user-images.githubusercontent.com/1937582/229906371-dc791ae4-e7f9-4139-8f01-4f1f46e0785d.png">
 
So, it makes sense to me at this point to support `label` within the `Toggle` itself, which is a lighter and simpler version of the `FormToggle`.
<img width="320" alt="image" src="https://user-images.githubusercontent.com/1937582/229906660-b5d11ee4-8078-4574-b047-209641d4c0c6.png">

ranquild replied:
If you need a `Toggle` with a label that is not within the form, you could do 

```
<FormField title={t`Auto-apply filters`} ...>
  <Toggle ... />
</FormField>
```

I don't think we should add another way of adding a label to a toggle/checkbox, wdyt?

ranquild replied:
Furthermore, I don't think we should be including `FormField` within the component itself, so I'd rather use the existing approach and use `FormField` + `Toggle` on the consumer side

WiNloSt replied:
Using `FormField` sounds good!

But, I saw that we have different kinds of styling labels too (see screenshots in the original comment). A more straightforward way of working that out is to add a styled file like `ChartSettingsWidget.tsx` here.
https://github.com/metabase/metabase/blob/6e03c5a663805852793b0e9da448259670ff12f4/frontend/src/metabase/visualizations/components/ChartSettingsWidget.tsx#L62-L66

<img width="323" alt="image" src="https://user-images.githubusercontent.com/1937582/230068922-4d039421-ade2-49c7-8d3c-464758ec3124.png">

The second approach could be to add a `labelVariant` (or any other better name) prop to `FormField`, so that we could standardize this design (which I think should even replace the ones we used in admin settings.

I'm in favor of adding a prop to `FormField` for different labels, what do you think?

WiNloSt commented on frontend/src/metabase/css/components/buttons.css:
```diff
@@ -22,7 +22,9 @@
   transition: all 200ms linear;
-  transition-property: color, border-color, background-color;
+  /* XXX: `visibility` and `opacity` are needed otherwise it won't transition when users click on the apply button, since it's essentially the hover state */
+  transition-property: color, border-color, background-color, visibility,
```
`visibility` and `opacity` are needed otherwise it won't transition and would disappear immediately when users click on the apply button, since it's essentially the hover state

ranquild replied:
If we need `visibility` transition for the `Apply` button, I believe we should add it in a styled component for the apply button and not in global css. Thoughts?

WiNloSt replied:
I could, but don't you think we should make all buttons support fade in and out? It feels more something that should be changed in the core component than a one-off thing.

This feels similar to a lot of changes I've seen done in visualization settings where we ended up changing the style of the component than adding styles to the `.styled.*` file.

What do you think?

WiNloSt commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:
```diff
@@ -36,6 +36,7 @@ import type {
 } from "metabase-types/types/Parameter";
 import type { Series } from "metabase-types/types/Visualization";
 
+import { DASHBOARD_SLOW_TIMEOUT } from "metabase/dashboard/constants";
```
standardize the timeouts, so they're easier to maintain

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:
```diff
@@ -344,4 +348,4 @@ class Dashboard extends Component {
 }
 
-export default DashboardControls(Dashboard);
+export default _.compose(DashboardControls, withAutoApplyFilters)(Dashboard);
```
A lot of logic are implemented using hooks, so I need to use a functional component, but `<Dashboard />` is a class component, so it's easier at this point to use HoC than to refactor the `<Dashboard />` to a functional component.

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.jsx:
```diff
@@ -90,11 +90,11 @@ export const ParametersAndCardsContainer = styled.div`
 
 export const ParametersWidgetContainer = styled(FullWidthContainer)`
-  align-items: flex-start;
+  align-items: baseline;
```
This helps align the parameters with the apply button.
<img width="742" alt="image" src="https://user-images.githubusercontent.com/1937582/229907891-a0650ed4-1181-45b1-b36e-8bd7f1eb3f57.png">

WiNloSt commented on frontend/src/metabase/dashboard/components/DashboardInfoSidebar.tsx:
```diff
@@ -47,17 +50,28 @@ const DashboardInfoSidebar = ({
-    async (description: string) => {
-      await setDashboardAttribute("description", description);
+    (description: string) => {
+      setDashboardAttribute("description", description);
```
this isn't an async function so we don't need to `await` it
https://github.com/metabase/metabase/blob/3fb962b9b816112a33f94079b63abe862dc7fe3f/frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx#L192-L197

ranquild replied:
I think it waits for action to be dispatched and proceed by the redcuers? Not sure if this change affects something or not

WiNloSt replied:
I'm pretty sure it doesn't wait for anything asynchronous. The referenced function isn't `async` so it doesn't matter what's even defined inside the function.

This in turn calls https://github.com/metabase/metabase/blob/734c39090455f64826e824ab59e51a76b999f649/frontend/src/metabase/dashboard/actions/core.js#L14

which is a sync action anyway. So we're getting the same result even if https://github.com/metabase/metabase/blob/3fb962b9b816112a33f94079b63abe862dc7fe3f/frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx#L192-L197 is to be an `async` function.

i.e. without an `await` this should already wait for the reducer to update all the values since it's a sync action.

my reason for the change is to avoid confusion. Async actions kind of imply some API calls would happen down the line. But there's none in this case.

WiNloSt commented on frontend/src/metabase/dashboard/components/DashboardInfoSidebar.tsx:
```diff
@@ -47,17 +50,28 @@ const DashboardInfoSidebar = ({
+    [dashboard.id, saveDashboardAndCards, setDashboardAttribute],
+  );
+
+  const handleToggleAutoApplyFilters = useCallback(
```
This function will also be used as another hook's dependency, so I need to memoize it.
https://github.com/metabase/metabase/blob/3fb962b9b816112a33f94079b63abe862dc7fe3f/frontend/src/metabase/core/components/Toggle/Toggle.tsx#L23-L28

WiNloSt commented on frontend/src/metabase/dashboard/constants.js:
```diff
@@ -6,3 +6,5 @@ export const SIDEBAR_NAME = {
   info: "info",
 };
+
+export const DASHBOARD_SLOW_TIMEOUT = 15 * 1000;
```
standardize the timeouts, so they're easier to maintain

WiNloSt commented on frontend/src/metabase/query_builder/components/VisualizationError.jsx:
```diff
@@ -9,6 +9,7 @@ import cx from "classnames";
 import MetabaseSettings from "metabase/lib/settings";
 import ErrorMessage from "metabase/components/ErrorMessage";
 import ErrorDetails from "metabase/components/ErrorDetails/ErrorDetails";
+import { VISUALIZATION_SLOW_TIMEOUT } from "../constants";
```
standardize the timeouts, so they're easier to maintain

WiNloSt commented on frontend/src/metabase/query_builder/constants.js:
```diff
@@ -24,3 +24,5 @@ export const SIDEBAR_SIZES = {
   TIMELINE: 300,
 };
+
+export const VISUALIZATION_SLOW_TIMEOUT = 15 * 1000;
```
standardize the timeouts, so they're easier to maintain

WiNloSt commented on frontend/src/metabase/parameters/components/ParametersList.jsx:
```diff
@@ -48,7 +48,7 @@ function ParametersList({
 
   setParameterValue,
   setParameterIndex,
-  removeParameter,
```
This is unused

WiNloSt commented on frontend/src/metabase/parameters/components/ParameterValueWidget.jsx:
```diff
@@ -154,6 +154,8 @@ class ParameterValueWidget extends Component {
               className={cx(S.parameter, className, {
                 [S.selected]: hasValue,
               })}
+              role="button"
+              aria-label={hasValue ? value : placeholder}
```
You can read the reason described in https://github.com/metabase/metabase/pull/29569/files#diff-124fa4dcae5da179c159091f4b0d0e3046144da58ab6ea3efa6741f69e18c2f6R91, why I made it a button.

TODO: Fix the link above.

ranquild replied:
I think the parameter value cannot be used as a label otherwise it would change every time the value changes. Could this change be reverted?

WiNloSt commented on frontend/src/metabase/dashboard/hoc/withAutoApplyFilters.tsx:
```diff
@@ -0,0 +1,109 @@
+    // the logic inside <Dashboard />'s `componentDidUpdate` will be triggered twice,
+    // causing the dashboard cards to be fetched twice, when there'are parameters
+    // in the query params.
+    const [state, setState] = useState<{
```
I need to control both `parameterValues`, and `dashboard` passed to `<Dashboard />` to be changed at the same time, otherwise the logic inside <Dashboard />'s `componentDidUpdate` will be triggered twice, causing the dashboard cards to be fetched twice, when there are parameters in the query params.

WiNloSt committed:
Remove puzzle driven development comments

WiNloSt commented on frontend/src/metabase/dashboard/hoc/withAutoApplyFilters.tsx:
```diff
@@ -0,0 +1,113 @@
+import { ApplyButton } from "./withAutoApplyFilters.styled";
+
+// XXX: Use similar type as in https://github.com/metabase/metabase/blob/c56d0872ac2a2cf44a827acfa1af5e8b76dcaa7d/frontend/src/metabase-lib/parameters/utils/cards.ts#L17
+type ParameterValues = Record<string, any>;
```
Use similar type as in https://github.com/metabase/metabase/blob/c56d0872ac2a2cf44a827acfa1af5e8b76dcaa7d/frontend/src/metabase-lib/parameters/utils/cards.ts#L17

ranquild commented on frontend/src/metabase/dashboard/hoc/withAutoApplyFilters.tsx:
```diff
@@ -0,0 +1,110 @@
+  }) => void;
+}
+
+export default function withAutoApplyFilters<T>(
```
It seems that this approach goes against the current one where `redux` is used to store dashboard-wide data. Instead of introducing another approach, could you move it to `redux` instead, alongside other dashboard reducers, actions, and selectors?

WiNloSt replied:
That's great advice. I haven't thought of that initially. I'll do that 👍

WiNloSt committed:
Move logic from HoC to redux 1/x

WiNloSt committed:
Fix x-ray dashboard parameters not working

WiNloSt committed:
Fix unit tests

WiNloSt committed:
Add E2E tests to public dashboard sharing

WiNloSt committed:
Add signed embedding E2E tests

WiNloSt commented on frontend/src/metabase/dashboard/components/DashboardInfoSidebar.tsx:
```diff
@@ -82,6 +97,20 @@ const DashboardInfoSidebar = ({
+          noMargin
+        >
+          <Toggle
+            aria-label={t`Auto-apply filters`}
```
I couldn't find a nice way to use `FormField`'s `title` for its children's `aria-label`. I notice it accepts `htmlFor`, but that includes the error message as well. 
https://github.com/metabase/metabase/blob/6476e38ac0b33fcc3a1ac069fd97b01347414441/frontend/src/metabase/core/components/FormField/FormField.tsx#L73-L80
 I don't want to be touching that part yet, but I believe the input label should be static and not depend on the state of the input. i.e. it should not change when the input has an error.

WiNloSt replied:
If I plainly remove `aria-label` from the `<Toggle />` [the test would simply fail](https://github.com/metabase/metabase/blob/734c39090455f64826e824ab59e51a76b999f649/e2e/test/scenarios/dashboard-filters/old-parameters.cy.spec.js#L88).
<img width="353" alt="image" src="https://user-images.githubusercontent.com/1937582/232475523-9734f8eb-6d3b-4907-a5ca-7a4f79e9077e.png">

I need to either
1. Add `aria-label` to the `<Toggle />` or
2. Set `id` to `<Toggle />` and pass that id to `<FormField />`'s `htmlFor`
    ![image](https://user-images.githubusercontent.com/1937582/232476070-175b7303-4715-4b2b-83ae-f1d5274d2bd7.png)


I think using `htmlFor` is preferable since we do not need to hard-code `aria-label` and we use an actual label from the DOM instead. What do you think?

WiNloSt commented on frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.styled.tsx:
```diff
@@ -86,6 +86,11 @@ const footerVariantStyles = {
   `,
 };
 
+export const ParametersWidgetContainer = styled.div`
+  display: flex;
+  align-items: baseline;
+`;
```
Matches the apply button styles on normal app dashboards. See `frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.jsx`

WiNloSt commented on frontend/src/metabase/dashboard/hoc/DashboardControls.jsx:
```diff
@@ -213,6 +216,29 @@ export default ComposedComponent =>
+          setParameterValues(draftParameterValues);
+        };
+
+        const applyFilterButton = !dashboard?.auto_apply_filters ? (
```
I opt to pass the whole `applyFilterButton` rather than passing a bunch of individual props and construct the button inside the view to keep related logic in the same place.

This component `DashboardControls` is shared among public dashboards and app dashboards.

ranquild replied:
Could you explain why decided to use this approach instead of rendering the button directly in the place it's supposed to be and "connect"-ing the button to not pass all the required props? This approach with passing ReactNode's around is not something we almost never do

ranquild replied:
You don't need to pass new props. You should be able to call `connect` or `useSelector` / `useDispatch` at the very bottom. WDYT?

WiNloSt commented on frontend/src/metabase/dashboard/reducers.js:
```diff
@@ -278,7 +278,13 @@ const parameterValues = handleActions(
     [SET_PARAMETER_VALUE]: {
-      next: (state, { payload: { id, value } }) => assoc(state, id, value),
+      next: (state, { payload: { id, value, isDraft } }) => {
+        if (!isDraft) {
```
as mentioned earlier `SET_PARAMETER_VALUE` will have a new payload attribute `isDraft` which determines which state we should set the parameter value to.

WiNloSt commented on frontend/src/metabase/dashboard/reducers.js:
```diff
@@ -294,6 +300,32 @@ const parameterValues = handleActions(
+    [Dashboards.actionTypes.UPDATE]: {
+      next: (state, { payload }) => {
+        if (payload.dashboard.auto_apply_filters) {
+          return null;
```
Currently, there's a bug that whenever we update a dashboard, the `parameterValues` will get reset. After that's fixed. I will modify this reducer so that if we update the filter while  auto-apply filters is disabled and toggle the auto-apply filters to true. We should copy the value inside `draftParameterValues` over to `parameterValues` so users don't lose their modified filters.

But to avoid confusion at the moment, and make things simple, I just clear the `draftParameterValues` so the behavior matches what it is at the moment (that the filters will be clear when the dashboard is updated).

WiNloSt commented on frontend/src/metabase/dashboard/reducers.unit.spec.js:
```diff
@@ -31,6 +31,7 @@ describe("dashboard reducers", () => {
         loadingStatus: "idle",
       },
       parameterValues: {},
+      draftParameterValues: null,
```
I opt for `null` rather than an empty object so that in the view I can simply use this `draftParameterValues ?? parameterValues`

ranquild replied:
Feels weird to have different behavior for `parameterValues` and draft ones, I'd use `{}` and then `_.isEmpty(draftParameterValues) ? parameterValues : draftParameterValues`

WiNloSt commented on frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.tsx:
```diff
@@ -128,19 +136,20 @@ function EmbedFrame({
                   parameters={getValuePopulatedParameters(
                     parameters,
-                    parameterValues,
+                    draftParameterValues ?? parameterValues,
```
Repeat the same change I made for app dashboards, but this is for public dashboards and signed dashboard emdeddings.

WiNloSt commented on frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:
```diff
@@ -25,7 +25,7 @@ describe("EmbedModalContent", () => {
   it("should render parameters", () => {
     const parameters = [
-      { name: "My param", slug: "my_param", type: "category" },
+      { name: "My param", slug: "my_param", type: "string/=" },
```
This doesn't make any difference to the test result, but I just couldn't find a way to set the parameter type to `category` at the moment.

Looking at `PARAMETER_OPERATOR_TYPES` you can see that there's no way to get a parameter with `category` type, unless I mistook this https://github.com/metabase/metabase/blob/6476e38ac0b33fcc3a1ac069fd97b01347414441/frontend/src/metabase-lib/parameters/constants.ts#L3

WiNloSt replied:
That makes sense

ranquild replied:
Could we omit this change? I think it's important to test `category` type. Maybe the root cause is that `Parameter` in `metabase-types/api` has a wrong type for `type` field. It should be `string` like it is in `metabase-types/types`. If that's the case here, let's fix the type in `api`

WiNloSt replied:
I can revert this change, at first I thought `category` was impossible, but you have stated it is possible.

WiNloSt commented on frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:
```diff
@@ -78,8 +78,14 @@ describe("EmbedModalContent", () => {
     openEmbedModal();
-    expect(screen.getByText("My param")).toBeInTheDocument();
-    expect(screen.getByLabelText("My param")).toHaveTextContent("Locked");
+    const parametersSection = screen
```
This fixes the signed dashboard embeddings e.g. in this screenshot
<img width="401" alt="image" src="https://user-images.githubusercontent.com/1937582/231563092-659dec93-4b63-4be1-a71d-d2e8f93142a5.png">

both dropdowns with a display label **Source w/ default** in both sections have the same label semantically, so `getByLabelText(label)` would find 2 elements.

This is the change introduced in this PR to make the parameter widget more accessible and testable. See the file `frontend/src/metabase/parameters/components/ParameterValueWidget.jsx`

WiNloSt commented on frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx:
```diff
@@ -78,8 +78,14 @@ describe("EmbedModalContent", () => {
-    expect(screen.getByLabelText("My param")).toHaveTextContent("Locked");
+    const parametersSection = screen
+      .getByRole("heading", { name: "Parameters" })
+      // eslint-disable-next-line testing-library/no-node-access
```
This feels a bit awkward but I could find we disable this ESLint rule is some other places as well. So I think it might be fine 🤷

ranquild replied:
If you need to differentiate between these sections you can use `role="section"` with `aria-label` for the section(s)

WiNloSt replied:
Yeah, this is the problem I talked about in our team sync that makes me want to create an RFC for section accessibility.  I think your solution might work.

WiNloSt commented on src/metabase/api/public.clj:
```diff
@@ -204,7 +204,7 @@
   [& conditions]
   {:pre [(even? (count conditions))]}
-  (-> (api/check-404 (apply t2/select-one [Dashboard :name :description :id :parameters], :archived false, conditions))
+  (-> (api/check-404 (apply t2/select-one [Dashboard :name :description :id :parameters :auto_apply_filters], :archived false, conditions))
```
Without this change, public dashboards will not know if they need to show the **Apply** button if `auto_apply_filters` is set to `false`. I also [confirmed with the product people](https://metaboat.slack.com/archives/C04DAKFDTCZ/p1681238761991669?thread_ts=1681233424.166269&cid=C04DAKFDTCZ).

WiNloSt commented on test/metabase/api/embed_test.clj:
```diff
@@ -149,7 +149,7 @@
 
 (def successful-dashboard-info
-  {:description nil, :parameters [], :ordered_cards [], :param_fields nil})
+  {:auto_apply_filters true, :description nil, :parameters [], :ordered_cards [], :param_fields nil})
```
Fix failed unit tests from modifying `src/metabase/api/public.clj`

ranquild commented on frontend/src/metabase/core/components/FormField/FormField.styled.tsx:
```diff
@@ -24,13 +25,22 @@ export const FieldCaption = styled.div<FormCaptionProps>`
 
 export interface FieldLabelProps {
   hasError: boolean;
+  titleStyle?: "small" | "large";
```
Unless there was an explicit design call to change this (1) for this specific place and make this conditional so we have 2 different labels in the app (2) we should not do changes like this.

WiNloSt replied:
@mazameli can you confirm this is a new style? I've seen this in a number of places e.g. visualization settings.
<img width="637" alt="Screenshot 2023-04-13 at 11 27 16 AM" src="https://user-images.githubusercontent.com/1937582/231731792-f76e2643-7137-45d0-8d13-17665431007b.png">

The only place I think I've seen the old style is in the admin settings forms.

ranquild reviewed (changes requested):
@WiNloSt Some decisions made in this PR make sense to me while others require discussion. To not block your work for a lot of time, could you please extract the following changes to separate PRs:
- BE changes
- FE - refactoring with constants + `noMargin` prop for the `FormField`

We could discuss this separately but I believe that we should drop button changes + `titleStyle` introduced in this PR.

Once 2 PRs above are extracted, please update the base branch in this PR and we will discuss parameter changes. I think that adding `draftParameterValues` makes sense, however I'd like to iterate on how the apply button is implemented so I'd rather talk about that when the PRs above are created. Thanks!

WiNloSt commented on frontend/src/metabase/core/components/Button/Button.styled.tsx:
```diff
@@ -11,6 +11,12 @@ export const ButtonRoot = styled.button<ButtonRootProps>`
+  &:hover {
+    /* XXX: standardize button transition, it wouldn't make much sense to only extend this in a single place, since transition should be a part of the design system. */
+    transition-property: color, border-color, background-color, opacity,
+      visibility;
```
I want to hear @mazameli 's opinion on this. I think we've had a couple of similar discussions about whether a certain style should be scoped to a specific UI components, or should that be baked into the code components.

From a design's point of view, I couldn't think of a way to want to only have a few buttons that would fade in and out while most of the other buttons instantly appear and disappear since that creates inconsistency in the design.

Let me know if fading in and out buttons should only be scoped to the **Apply** button specifically for this project.

qnkhuat reviewed (approved):
BE LGTM

WiNloSt commented:
@ranquild sounds good extracting what you've mentioned to another PR, tbh I can actually remove the constant refactor, that's for the 2nd milestone, and I think that got mixed up when splitting the PR.

WiNloSt committed:
Make apply filter button works on embeddings

WiNloSt committed:
Fix the apply button transition not working when clicked

Since there's another CSS applied when the button is hovered.

WiNloSt committed:
Able to toggle auto-apply dashboard filters

Make FormField be able to show title with new style

ranquild reviewed (commented):
We connect components to the store to avoid props drilling and not pass `ReactNode`s around. @WiNloSt Would it be possible to use `ApplyButton` directly where it's rendered and `connect` it to the store to avoid props drilling? And remove passing `applyButton` node. If not, could you explain what prevents that?

WiNloSt commented:
@ranquild I've replied to https://github.com/metabase/metabase/pull/29569#pullrequestreview-1383745286 with https://github.com/metabase/metabase/pull/29569/files#r1165464555.

That comment was published later since I was not realized it was pending publishing.

If you think an alternative is still preferred I'll make a change, but now I won't modify anything just to not waste another CI cycle.

Let me know what you think.

ranquild commented:
@WiNloSt I still don't understand why we pass `ReactNode` for the `applyButton` and cannot render it directly and `connect` to the store (e.g. pass needed props and dispatch actions). I think this change wasn't explained in the comments. You said that you want to avoid props drilling but no props drilling is required when connecting to the redux store.

WiNloSt commented on frontend/src/metabase-types/api/dashboard.ts:
```diff
@@ -29,6 +29,7 @@ export interface Dashboard {
     last_name: string;
     timestamp: string;
   };
+  auto_apply_filters: boolean;
```
Good catch, making it `false` for X-ray would make it a lot simpler, so we don't even need to introduce a new util like this https://github.com/metabase/metabase/pull/29569/files#r1168505420, and I could revert that change to just a typical boolean check. This is one of the examples that I think booleans with 3 states are confusing.

WiNloSt committed:
Reviews: Fix x-ray dashboard actions

Remove the need of 3-states boolean in `setParameterValue` action for
x-ray dashboards. This fixes so that BE sends x-ray dashboards with
`auto_apply_filters: true`.

WiNloSt committed:
Review: remove duplicate `aria-label`

WiNloSt committed:
Review: standadize `draftParameterValues` default values

WiNloSt committed:
Review: Fix `aria-label` to not change when filter values change

WiNloSt commented on frontend/src/metabase/dashboard/actions/parameters.js:
```diff
@@ -175,8 +175,10 @@ export const setParameterFilteringParameters = createThunkAction(
   SET_PARAMETER_VALUE,
-  (parameterId, value) => (dispatch, getState) => {
-    return { id: parameterId, value };
+  (parameterId, value) => (_dispatch, getState) => {
```
Fixed

WiNloSt commented on frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.tsx:
```diff
@@ -108,6 +122,16 @@ function EmbedFrame({
 
   const hasHeader = Boolean(finalName || hasParameters);
 
+  const applyFilterButton = shouldRenderApplyButton ? (
```
Fix passing `ReactNode` by constructing the button inside the view components instead.

I try to also make it as clear as possible which props are for auto-apply filters.

WiNloSt commented on frontend/src/metabase/dashboard/reducers.unit.spec.js:
```diff
@@ -31,6 +31,7 @@ describe("dashboard reducers", () => {
         loadingStatus: "idle",
       },
       parameterValues: {},
+      draftParameterValues: {},
```
Fixed default value on `draftParameterValues` so it looks the same as `parameterValues` to avoid confusion.

WiNloSt commented on frontend/src/metabase/parameters/components/ParameterValueWidget.jsx:
```diff
@@ -154,6 +154,8 @@ class ParameterValueWidget extends Component {
                 [S.selected]: hasValue,
               })}
+              role="button"
+              aria-label={placeholder}
```
Fixed so that we only use `placeholder` as a label not also `value` which could change every time users change filters.

WiNloSt commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -1079,12 +1079,13 @@
+                                     (format "%s#show=all" (:url root)))
+               :transient_filters  (:query-filter context)
+               :param_fields       (->> context :query-filter (filter-referenced-fields root))
+               :auto_apply_filters true))))
```
make sure there is `auto_apply_filters` in the API response for x-ray dashboards.

WiNloSt commented:
@ranquild I've addressed all your comments. Please take a look. Also, I bundled the change from https://github.com/metabase/metabase/pull/30071 (which only have 1 commit before the change) into this PR since it's small enough.

ranquild reviewed (commented):
There is props drilling that we should avoid. You can do it by using actions and selectors from redux. Here is the skeleton that might help:

```
// actions

const applyDraftParameterValues = createThunkAction("...", () => (dispatch, getState) => {
  const values = getDraftParameterValues(getState());
  dispatch(setParameterValues(values));
})

// selectors

const getHasAutoApplyFilters = createSelector(
  [getDashboardComplete], 
  (dashboard) => dashboard.auto_apply_filters,
);

const getHasUnappliedParameterValues = createSelector(
  ...
);

// button component

const mapStateToProps = state => ({
  hasAutoApplyFilters: getHasAutoApplyFilters(state),
  hasUnappliedParameterValues: getHasUnappliedParameterValues(state),
});

const mapDispatchToProps = {
  onApplyFilters: applyDraftParameterValues,
};

const FilterApplyButton = ({
  hasAutoApplyFilters,
  hasUnappliedParameterValues,
  onApplyFilters,
}) => {
  if (!hasAutoApplyFilters) {
    return null;
  }

  return (
    <ApplyButton
      isVisible={hasUnappliedParameterValues}
      onClick={onApplyFilters}
    />
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterApplyButton);

// Dashboard, EmbedFrame components
// no props needs to be passed

<FilterApplyButton />

```

Could you take a look at this? Thanks!

WiNloSt committed:
Review: avoid passing `ReactNode` as a prop

Also fix transition not working and apply button shift when adding a
filter.

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:
```diff
@@ -309,6 +317,7 @@ class Dashboard extends Component {
                     topNav={embedOptions?.top_nav}
                   >
                     {parametersWidget}
+                    <FilterApplyButton />
```
Use a dedicated component that connects to all the necessary data so we don't need to pass any prop to it.

WiNloSt commented on frontend/src/metabase/parameters/components/FilterApplyButton/FilterApplyButton.styled.ts:
```diff
@@ -0,0 +1,40 @@
+    transition-property: opacity, visibility;
+  }
+
+  &.${TRANSITION_CLASSNAMES_PREFIX}-enter {
```
A bunch of these declarations are for transitioning the button when mounting and unmounting the component using `react-transition-group`.

You can test that when toggling the Auto-apply filters toggle on and off there should be a transition too.

WiNloSt commented on frontend/src/metabase/parameters/components/FilterApplyButton/FilterApplyButton.styled.ts:
```diff
@@ -0,0 +1,40 @@
+
+export const ApplyButton = styled(Button)<{ isVisible: boolean }>`
+  margin-left: auto;
+  margin-top: ${space(1)};
```
Previously I use `align-items: baseline` on the parent container, but that causes a slight shift when adding a filter. So I use fixed margin to make the apply button stay still.
![apply button shift](https://user-images.githubusercontent.com/1937582/232624252-262d857c-f47f-48fb-99fd-99608f6435d7.gif)

WiNloSt commented on frontend/src/metabase/parameters/components/FilterApplyButton/FilterApplyButton.tsx:
```diff
@@ -0,0 +1,60 @@
+} from "./FilterApplyButton.styled";
+
+export default function FilterApplyButton() {
+  const isAutoApplyFilters = useSelector(
```
Use new `react-redux` hooks

ranquild replied:
We should define selectors in the appropriate file where selectors are located for this feature. For `dashboard` feature folder it's `metabase/dashboard/selectors.js`.

The same with actions - we shouldn't (and we don't) define actions (thunks) inline. Could you move redux action code from `handleApplyFilters` to an action in `metabase/dashboard/actions` folder like in the code snippet I've shared before?

WiNloSt replied:
That's fair.

WiNloSt commented on frontend/src/metabase/parameters/components/FilterApplyButton/FilterApplyButton.tsx:
```diff
@@ -0,0 +1,60 @@
+  const applyButtonRef = useRef(null);
+
+  return (
+    <CSSTransition
```
Use `react-transition-group` to allow transitioning component when mounting and unmounting

WiNloSt replied:
@ranquild Just want to make sure we're on the same page about this.
1. Is `react-transition-group` discouraged in Metabase?
2. I haven't seen any doc saying it's outdated or deprecated, if we look at their updated date
    1. [`react-transition-group`](https://www.npmjs.com/package/react-transition-group) was updated 9 months ago vs 
    2. [`react-motion`](https://www.npmjs.com/package/react-motion) was updated 6 years ago
3. TBH, if we want to stay up to date, I'd rather use another modern library e.g. [`framer-motion`](https://www.framer.com/motion), but that's outside of the scope of this project.
4. `react-motion`'s [use case](https://github.com/chenglou/react-motion#what-does-this-library-try-to-solve) is for animating components with different curves that are not only limited to CSS properties, I feel it isn't worth trying to migrate over to `react-motion` for this PR's use-case which is just to animate mounting and unmounting components.

That being said, what do you think about leaving `react-transition-group` change as is? Or at least I could remove the use of `react-transition-group` and don't animate mounting and unmounting components if we don't really like to use `react-transition-group`

ranquild replied:
@WiNloSt would it be possible to do the transition with pure css in styled components like you did before? Are there any issues with that?

WiNloSt replied:
transitioning mounting and unmounting were never support with pure CSS without a lot of work. The previous version didn't fade apply button in and out on mounting and unmounting the apply button component.

We need to remove it from the dom entirely as discussed with Maz so when `auto_apply_filters` is `true` the apply button won't take any space.

Previously when the component is mounting it just appear on the DOM, so to animate mounting the component we need a mechanism which add class name to the component one after another in a different tick so CSS could transition from say opacity 0 to 1. This is the work that's handled by `CSSTransition` from `react-transition-group`. The same goes with unmounting the component, without some work to keep the component around for the whole amount of the animation, if we just conditionally render the component, it will just be removed from the DOM without a chance to be animated, this is the work that's also handled by these kinds of transition/animation libraries.

ranquild replied:
@WiNloSt Is it even possible that we need to animate the button when the auto_apply_filters is changed? It can only change in the edit mode (or while loading in milestone 2), there are no draft parameter values at this stage. I'd vote for a simple solution with a css transition like you did before. Would that be possible?

WiNloSt replied:
@ranquild Yeah pure CSS would be better, in this case, to keep things simple. There are cases, that we (might) need to animate unmounting and mounting the component that is when you have an unapplied filters so there's the apply button visible then you toggle the auto-apply filters back to on. This should remove the apply button out of the dom. I mean it'd be nice just to keep the experience consistent when removing them from the dom and fade them out, but if that begs more question I'm cool with only using pure CSS and not animating them on mount and unmount.

So I'll animate the apply button only when it is already on the DOM, no animation when unmounting or mounting (which is rare anyway). Sounds good?

ranquild replied:
Yep sounds great

WiNloSt commented on frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.styled.tsx:
```diff
@@ -86,6 +86,11 @@ const footerVariantStyles = {
   `,
 };
 
+export const ParametersWidgetContainer = styled.div`
```
the parameter list in all embeddings (including public sharing) uses a different container, so this style is to make it renders like `ParametersWidgetContainer` inside `frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.jsx`

WiNloSt commented on frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.tsx:
```diff
@@ -128,19 +134,22 @@ function EmbedFrame({
                   hideParameters={hide_parameters}
                 />
-              </div>
+                {dashboard && <FilterApplyButton />}
```
Since `EmbedFrame` not only renders dashboards but questions as well. And in those cases `FilterApplyButton` will crash since it expects a dashboard to exist.

WiNloSt commented:
@ranquild I just got it when you provide the snippet above that we can create a component like that. To be honest, I haven't thought of that since, I think it's been too long since I've been creating new components and connecting them to redux 😅 

In this iteration, I added a fix that transitions the apply button when it's mounting and unmounting (via toggling the auto-apply filters toggle). Previously it will appear and disappear instantly.

qnkhuat reviewed (approved):
Thanks for adding a test 💪

ranquild commented on frontend/src/metabase/dashboard/actions/parameters.js:
```diff
@@ -175,8 +175,10 @@ export const setParameterFilteringParameters = createThunkAction(
-    return { id: parameterId, value };
+  (parameterId, value) => (_dispatch, getState) => {
+    const isSettingDraftParameterValues = !getDashboard(getState())
+      .auto_apply_filters;
```
When you define a selector for `auto_apply_filters`, make sure to update this line and re-use it.

WiNloSt replied:
fixed

ranquild commented on e2e/test/scenarios/sharing/public.cy.spec.js:
```diff
@@ -200,5 +201,48 @@ describe("scenarios > public", () => {
+          .should("not.be.checked");
+      });
+
+      Object.entries(USERS).map(([userType, setUser]) =>
```
We have quite a lot of users to iterate over them in e2e like that. Rather than testing for all the users, is there anything specific you want to check for? To me it seems that this feature doesn't depend on user attributes/permissions so I'd just have a single test.

WiNloSt replied:
Fixed by only running a single test now since it's a public dashboard so doesn't need to be tested by different users.

ranquild commented on e2e/test/scenarios/embedding/embedding-linked-filters.cy.spec.js:
```diff
@@ -106,6 +107,84 @@ describe("scenarios > embedding > dashboard > linked filters (metabase#13639, me
+        });
+
+        visitDashboard(dashboard_id);
+        visitEmbeddedPage(payload);
```
Is it intended to visit the dashboard and then immediately visit the embedded page?

WiNloSt replied:
nice catch, it was probably my unintentionally paste from my clipboard

WiNloSt replied:
Fixed by removing `visitDashboard` that was unintentional.

ranquild commented on e2e/test/scenarios/embedding/embedding-linked-filters.cy.spec.js:
```diff
@@ -106,6 +107,84 @@ describe("scenarios > embedding > dashboard > linked filters (metabase#13639, me
+      });
+
+      cy.findByRole("heading", { name: nativeDashboardDetails.name });
+      cy.get(".Card").contains(nativeQuestionDetails.name);
```
There is `getDashboardCard` e2e helper to avoid using the css class

WiNloSt replied:
fixed

ranquild commented on e2e/test/scenarios/dashboard-filters/old-parameters.cy.spec.js:
```diff
@@ -63,14 +68,35 @@ describe("scenarios > dashboard > OLD parameters", () => {
+      // test prevent auto-apply filters in slow dashboards metabase#29267
+      openDashboardSidebar();
+
+      const dashboardParameters = cy.findByTestId(
```
There is `filterWidget` helper for this, please search how it's used in other tests

ranquild reviewed (commented):
Looks good however there some things we still need to address. We need to remove `react-transition-group` dependency and extract actions & selectors to appropriate files.

WiNloSt committed:
Move selectors and actions to their own file

WiNloSt committed:
Review: Don't animate (un)mount of FilterApplyButton

WiNloSt committed:
Address: Clean up eslint ignore on tests

WiNloSt committed:
Review: Only test 1 scenario in E2E

Since testing other scenarios are unnecessary since they're repeat the
same thing

WiNloSt committed:
Review: Use E2E helper to clean up test

WiNloSt committed:
Review: Reuse selector

WiNloSt committed:
Review: Use e2e helpers to clean up the test

WiNloSt committed:
Fix typo

WiNloSt commented on frontend/src/metabase/dashboard/selectors.js:
```diff
@@ -102,6 +102,20 @@ export const getDashboardComplete = createSelector(
+export const getDraftParameterValues = state =>
+  state.dashboard.draftParameterValues;
+export const getIsAutoApplyFilters = createSelector(
+  [getDashboard],
```
using `getDashboard` the same selector that's used in `frontend/src/metabase/dashboard/actions/parameters.js`

WiNloSt commented on frontend/src/metabase/public/components/widgets/AdvancedSettingsPane.jsx:
```diff
@@ -112,11 +113,17 @@ const AdvancedSettingsPane = ({
+const Section = ({ className, title, children }) => {
+  const sectionId = useUniqueId();
+  return (
+    <section
```
This change is used in `frontend/src/metabase/public/components/widgets/EmbedModalContent.unit.spec.tsx`

ranquild commented on frontend/src/metabase/parameters/components/FilterApplyButton/FilterApplyButton.tsx:
```diff
@@ -0,0 +1,38 @@
+    dispatch(applyDraftParameterValues());
+  }, [dispatch]);
+
+  const applyButtonRef = useRef(null);
```
nit: seems not used

WiNloSt replied:
Nice catch. It was from when I was using `react-transition-group`. I'll need to see why ESLint doesn't catch this.

WiNloSt replied:
Ok, I see that it's referenced, but it's not really used for anything that's why ESLint didn't catch it.

WiNloSt replied:
Fixed

ranquild reviewed (approved):
Great work!

uladzimirdev commented on e2e/test/scenarios/dashboard-filters/old-parameters.cy.spec.js:
```diff
@@ -63,14 +68,35 @@ describe("scenarios > dashboard > OLD parameters", () => {
+      rightSidebar().findByLabelText("Auto-apply filters").click();
+      dashboardParameters.button("Apply").should("not.exist");
+
+      cy.button("Category").click();
```
off top, but `cy.button` seems to be an antipattern in cypress
<img width="1015" alt="image" src="https://user-images.githubusercontent.com/125459446/233047595-b1f3ed50-cbd5-470a-b1c1-d64f06b84641.png">

WiNloSt replied:
That excerpt looks a bit different from this case. That one just abstracts `cy.get(selector).click()` but our `cy.button` does help with not having to type long `cy.findByRole("button", { name: buttonName })` every time.

https://github.com/metabase/metabase/blob/34b9ff849b080baeff5858f0aeab7eb35db93b0a/e2e/support/commands/ui/button.js#L1-L3

I think the use case is a bit different, though I personally don't mind typing `.findByRole("button", { name: xxx })` all the time since that's what we do with non-button elements as well. But @nemanjaglumac might be able to give more insight. As he suggests I use `cy.button`.
